[Fix #1337] Update the cop to raise error for insufficient descriptions.

akrox58 opened the pull request:
Per issue #1337, I tried to add a generic way to the ExampleWordingCop to call out insufficient descriptions. For now, I've used "works" and "throws errors" as that is a phrase I see used often as well. I didn't use a new cop since insufficient descriptions seemed to be along the same lines as "should not start with should or it" as they are both description related.

---

Before submitting the PR make sure the following are checked:

* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [X] Added tests.
* [X] Updated documentation.
* [X] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
* [X] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have modified an existing cop's configuration options:

* [X] Set `VersionChanged` in `config/default.yml` to the next major version.

ydah commented on lib/rubocop/cop/rspec/example_wording.rb:
```diff
@@ -29,14 +30,34 @@ module RSpec
 
         SHOULD_PREFIX = /\Ashould(?:n't)?\b/i.freeze
         IT_PREFIX     = /\Ait /i.freeze
+        INSUFFICIENT_EXAMPLES = ['works', 'throws error'].freeze
```
It would be better if it could be changed with a configurable option, as suggested in https://github.com/rubocop/rubocop-rspec/issues/1337#issuecomment-1196650696

akrox58 replied:
DisallowedExamples:
    - works
    
My concern with adding it as a configurable option is the possibility of how big the list can get 😅 Are we anticipating it to get really big eventually?

ydah replied:
I think each project would have different examples of settings you want to disallowed, so wouldn't it be better to allow each project to configure its own settings?

akrox58 replied:
ah gotcha!!! makes sense!!

bquorning commented on lib/rubocop/cop/rspec/example_wording.rb:
```diff
@@ -52,6 +73,9 @@ def on_block(node)
               add_wording_offense(description_node, MSG_SHOULD)
             elsif message.match?(IT_PREFIX)
               add_wording_offense(description_node, MSG_IT)
+            elsif INSUFFICIENT_EXAMPLES.include?(message.strip.gsub(/  +/, ' '))
```
```suggestion
            elsif INSUFFICIENT_EXAMPLES.include?(message.strip.squeeze(' '))
```

akrox58 replied:
Thank you for the correction!

pirj commented on spec/rubocop/cop/rspec/example_wording_spec.rb:
```diff
@@ -187,4 +187,29 @@
+
+  it 'flags an unclear description' do
+    expect_offense(<<-'RUBY')
+      it "throws error" \
+          ^^^^^^^^^^^^^^^ Your test should be more descriptive.
+        "with #{object}" do
```
Those should be examples of what the cop should ignore.
The full message is "throws error with a message". Why is this an offence?

It seems that:
 - we don't consider articles e.g. "throws an error"
 - we don't inspect complete messages, but just the first part of it

akrox58 replied:
Thanks for bringing up the article part! you are right that we don't consider articles! I went down a rabbit hole of trying to ignore the articles and strip the phrase to just ["throws error"] from ["throws error, throws an error"] etc... But then I realized there could be grammatical errors that get missed from it. so I'm a little confused at what to do. I removed the phrase for the time being though. 

as for #2, I thought include? checks for whole phrases. so we will ignore "throws an error with a message" but the cop will sound an alarm for "throws error". Is there something I'm missing?

ydah commented on lib/rubocop/cop/rspec/example_wording.rb:
```diff
@@ -29,11 +30,21 @@ module RSpec
       #   it 'does things' do
       #   end
+      #
+      # @example
```
It would be good to have a good case/bad case according to the setting values of the configuration options.

akrox58 replied:
I think I addressed it with my latest changes, but let me know if you were expecting something else!

ydah replied:
I think the part you added is an explanation of the case where 'it' is repeated. I thinking of the https://github.com/rubocop/rubocop-rspec/pull/1350#discussion_r938271855 changes. For your information.

ydah commented on lib/rubocop/cop/rspec/example_wording.rb:
```diff
@@ -112,6 +126,14 @@ def custom_transform
+        end
+
+        def clean_msg(message)
+          message.strip.squeeze(' ').downcase
```
If I format it with `downcase`, won't it always mismatch if there are uppercase letters in the `DisallowedExample` setting?

akrox58 replied:
good callout! we should not allow uppercase in the DisallowedExample setting! Updating!

akrox58 replied:
great callout. I've added a comment that the words in this config will not be treated as case sensitive!

ydah replied:
I think it looks good.
Also, would it look good to add the following test?

```ruby
context 'when `DisallowedExamples: Workz`' do
  let(:cop_config) { { 'DisallowedExamples' => ['Workz'] } }

  it 'flags an unclear description' do
    expect_offense(<<-'RUBY')
      it "workz" \
          ^^^^^^^^ Your test should be more descriptive.
        "with #{object}" do
      end
    RUBY
  end

  it 'flags an unclear description despite extra spaces' do
    expect_offense(<<-'RUBY')
      it "  workz    " \
          ^^^^^^^^^^^^^^ Your test should be more descriptive.
        "with #{object}" do
      end
    RUBY
  end

  it 'flags an unclear description despite uppercase and lowercase strings' do
    expect_offense(<<-'RUBY')
      it "WOrKz " \
          ^^^^^^^^^ Your test should be more descriptive.
        "with #{object}" do
      end
    RUBY
  end
end
```

akrox58 replied:
@ydah  I thought I saw a notification on updating the test cases! that was a good idea! have added that too! 
Appreciate your patience on this! 
PR is ready for review again 😅

ydah replied:
Oops, I thought I had added a comment about the proposed test, but it disappeared. Thank you so much!

akrox58 commented on lib/rubocop/cop/rspec/example_wording.rb:
```diff
@@ -21,19 +26,29 @@ module RSpec
       #   end
       #
-      # @example
+      # @example `DisallowedExamples: ['works']`
       #   # bad
       #   it 'it does things' do
       #   end
       #
       #   # good
       #   it 'does things' do
       #   end
+      #
+      # @example
+      #   # bad
+      #   it 'works' do
+      #   end
+      #
+      #   # good
+      #   it 'marks the task as done' do
+      #   end
```
Ahhh sorry typo :/

ydah reviewed (approved):
Looks good to me, thank you.

akrox58 commented:
Thanks @ydah! How do I get this merged? Do I need to wait for the other reviewers to approve?

pirj reviewed (changes requested):
I would expect the following to pass, but it fails, why?
```ruby
    expect_no_offenses(<<-'RUBY')
      it "works" \
        "totally fine" do
      end
    RUBY
```

akrox58 commented:
> I would expect the following to pass, but it fails, why?
> 
> ```ruby
>     expect_no_offenses(<<-'RUBY')
>       it "works" \
>         "totally fine" do
>       end
>     RUBY
> ```

Looks like I didn't handle that edge case! updated the check to check for the entire description instead, and added some more test cases! Thanks for the catch!

pirj reviewed (approved):
I've checked against https://github.com/pirj/real-world-rspec, no errors, and the cop "just works" 😄 

```
...
/Users/pirj/source/real-world-rspec/sharetribe/spec/models/person_spec.rb:338:9: C: [Correctable] RSpec/ExampleWording: Your test should be more descriptive.
    it 'works' do
        ^^^^^
/Users/pirj/source/real-world-rspec/solidus/core/spec/lib/spree/core/testing_support/factories/address_factory_spec.rb:41:11: C: [Correctable] RSpec/ExampleWording: Your test should be more descriptive.
      it 'works' do
          ^^^^^

57497 files inspected, 273 offenses detected, 273 offenses autocorrectable
```

Looks great, thank you for the contribution!

akrox58 commented on config/default.yml:
```diff
@@ -368,8 +368,10 @@ RSpec/ExampleWording:
     have: has
     HAVE: HAS
   IgnoredWords: []
+  DisallowedExamples:
+    - works
```
@pirj Is there something I would need to do to address this?

pirj replied:
No, it is fine. Just thinking out loud to get an opinion of maintainers.
For me, this change is totally fine as is.

bquorning commented:
I have renamed some of the required CI checks, which unfortunately means that you will need to rebase this branch to make CI pass.

pirj commented:
Rebased and green @bquorning @Darhazer

bquorning commented on CHANGELOG.md:
```diff
@@ -38,6 +38,7 @@
 * Fix a false positive for `RSpec/Capybara/SpecificMatcher` when pseudo-classes. ([@ydah][])
 * Fix a false negative for `RSpec/SubjectStub` when the subject is declared with the `subject!` method and called by name. ([@eikes][])
 * Support `Array.new(n)` on `RSpec/FactoryBot/CreateList` cop. ([@r7kamura][])
+* Update `RSpec/ExampleWording` cop to raise error for insufficient descriptions. ([@akrox58][])
```
Could you move this up to the `(Unreleased)` section of the Changelog, please?

bquorning commented on lib/rubocop/cop/rspec/example_wording.rb:
```diff
@@ -113,6 +135,15 @@ def custom_transform
+
+        def insufficient_examples
+          examples = cop_config.fetch('DisallowedExamples', [])
+          examples.map! { |example| preprocess(example) }
```
Won't this line allocate (at least) 3 String objects on each `on_block` call? I don’t know if it would be possible to memoize this at class level.

pirj replied:
Since, in theory, different `DisallowedExamples` can be configured in different directories, it would be incorrect to cache this at the class level.
I've made the change to cache it in an instance variable, as I guess the same instance will be used for a single investigation.
We also need to [nullify the instance variable in `on_new_investigation`](https://docs.rubocop.org/rubocop-rspec/upgrade_to_version_2.html#always-call-super-from-on_new_investigation-in-your-cops). Is it worth it?

bquorning replied:
No, probably not worth it. Thanks for investigating.

bquorning reviewed (commented):
> This cop will also look for insufficient examples and call them out.

> Use the DisallowedExamples setting to prevent unclear or insufficient descriptions.

> `MSG_INSUFFICIENT_DESCRIPTION = 'Your test should be more descriptive.'`

I think we use a more clear terminology. Internally in rspec-core, the first argument to `it` is called `desc`. So perhaps

> This cop will also look for insufficient example descriptions and call them out.

> Use the DisallowedExampleDescriptions setting to prevent unclear or insufficient example descriptions.

> `MSG_INSUFFICIENT_DESCRIPTION = 'Your example description is insufficient.'`

pirj commented:
Thank you for the contribution and your time, @akrox58 !
